When BLE connection failed in android, application getting runtime abort due to pending ClassNotFoundException on "io.github.gedgygedgy.rust.future.FutureException"

[logcat-filtered.txt](https://github.com/deviceplug/btleplug/files/11713944/logcat-filtered.txt)

[app-debug.apk.zip](https://github.com/deviceplug/btleplug/files/11713953/app-debug.apk.zip)


**Describe the bug**
When BLE connection is failed in android, runtime abort of application is observed due to pending class not found exception.
The exception is java.lang.ClassNotFoundException and class it is unable to find is "io.github.gedgygedgy.rust.future.FutureException".
Wrote an android application which will load rust native library and print the services and characteristics of nearby BLE peripheral devices. In the Rust library, I have used btleplug/examples/discover_adapters_peripherals.rs and called the function from JNI method.

**Expected behavior**
When BLE connection is timedout or rejected, error should be printed and continued to prepare connection request to next peripheral in the scanned list. The android application should not be stopped.

**Actual behavior**
But the application is getting runtime abort due to pending ClassNotFoundException on "io.github.gedgygedgy.rust.future.FutureException". But that class is part of dependent java library to droidplug-debug.aar library and droidplug-debug.aar library is added as build time dependency to the android application.
No issue observed when BLE connection is successful.

**Additional context**
logcat log is attached and apk also attached for reference.
snippet of the crash signature is pasted below.

2023-06-11 10:27:46.853 21775-21798 BluetoothGatt           com.example.btleplugex               D  onClientConnectionState() - status=133 clientIf=9 device=DA:7C:35:35:F2:DF
2023-06-11 10:27:46.857 21775-21807 btleplugex-jni          com.example.btleplugex               E  btleplugex::discover_adapters_peripherals: Error connecting to peripheral, skipping: Other(JavaException)
2023-06-11 10:27:46.858 21775-21807 mple.btleplugex         com.example.btleplugex               A  java_vm_ext.cc:594] JNI DETECTED ERROR IN APPLICATION: JNI GetMethodID called with pending exception java.lang.**ClassNotFoundException**: Didn't find class **"io.github.gedgygedgy.rust.future.FutureException"** on path: DexPathList[[directory "."],nativeLibraryDirectories=[/system/lib64, /system_ext/lib64, /system/lib64, /system_ext/lib64]]
                                                                                                    java_vm_ext.cc:594]   at java.lang.Class dalvik.system.BaseDexClassLoader.findClass(java.lang.String) (BaseDexClassLoader.java:259)
                                                                                                    java_vm_ext.cc:594]   at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String, boolean) (ClassLoader.java:379)
                                                                                                    java_vm_ext.cc:594]   at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String) (ClassLoader.java:312)
                                                                                                    java_vm_ext.cc:594]
                                                                                                    java_vm_ext.cc:594]     in call to GetMethodID
2023-06-11 10:27:46.985 21775-21807 mple.btleplugex         com.example.btleplugex               A  runtime.cc:682] Runtime aborting...
                                                                                                    runtime.cc:682] Dumping all threads without mutator lock held
.....

Found another crash signature

got this when the peripheral device is powered down while subscribing to notify characteristic.

`2023-06-19 23:39:02.928 25045-25706 btleplugex-jni          com.example.btleplugex               I  btleplugex::discover_adapters_peripherals: Subscribing to characteristic Characteristic { uuid: 00002a37-0000-1000-8000-00805f9b34fb, service_uuid: 0000180d-0000-1000-8000-00805f9b34fb, properties: NOTIFY }
2023-06-19 23:39:07.930 25045-25706 BluetoothGatt           com.example.btleplugex               D  setCharacteristicNotification() - uuid: 00002a37-0000-1000-8000-00805f9b34fb enable: true
2023-06-19 23:39:07.940 25045-25706 System.out              com.example.btleplugex               I  Uma->setCommandCallback() callback: com.nonpolynomial.btleplug.android.impl.Peripheral$6@96b95c1
2023-06-19 23:39:09.839 25045-25062 System.out              com.example.btleplugex               I  Uma->wakeWithThrowable(), result: java.lang.RuntimeException: Unable to write descriptor
2023-06-19 23:39:09.841 25045-25065 BluetoothGatt           com.example.btleplugex               D  onClientConnectionState() - status=0 clientIf=15 device=FB:3C:36:23:FA:AE
2023-06-19 23:39:09.843 25045-25706 btleplugex-jni          com.example.btleplugex               D  jni::wrapper::jnienv: exception found, returning error
2023-06-19 23:39:09.843 25045-25706 btleplugex-jni          com.example.btleplugex               D  jni_utils::exceptions: Uma-> TryCatchResult catch
2023-06-19 23:39:09.843 25045-25706 btleplugex-jni          com.example.btleplugex               D  jni_utils::exceptions: Uma-> TryCatchResult step 4, received Ok(Err(Error::JavaException)) type
2023-06-19 23:39:09.843 25045-25706 btleplugex-jni          com.example.btleplugex               D  btleplug::droidplug::peripheral: Uma->unknown exception, re-throwing
2023-06-19 23:39:09.844 25045-25706 btleplugex-jni          com.example.btleplugex               E  btleplugex: discover() returned error: Other(JavaException)
--------- beginning of crash
2023-06-19 23:39:09.844 25045-25706 btleplugex-jni          com.example.btleplugex               D  btleplugex: exiting discover thread: ThreadId(30)
2023-06-19 23:39:09.846 25045-25706 AndroidRuntime          com.example.btleplugex               E  FATAL EXCEPTION: Thread-22
                                                                                                    Process: com.example.btleplugex, PID: 25045
                                                                                                    io.github.gedgygedgy.rust.future.FutureException: java.lang.**RuntimeException**: Unable to write descriptor
                                                                                                    	at io.github.gedgygedgy.rust.future.SimpleFuture.lambda$wakeWithThrowable$1(SimpleFuture.java:76)
                                                                                                    	at io.github.gedgygedgy.rust.future.SimpleFuture$$ExternalSyntheticLambda1.get(Unknown Source:2)
                                                                                                    Caused by: java.lang.RuntimeException: Unable to write descriptor
                                                                                                    	at com.nonpolynomial.btleplug.android.impl.Peripheral$6.lambda$onDescriptorWrite$0$com-nonpolynomial-btleplug-android-impl-Peripheral$6(Peripheral.java:246)
                                                                                                    	at com.nonpolynomial.btleplug.android.impl.Peripheral$6$$ExternalSyntheticLambda0.run(Unknown Source:10)
                                                                                                    	at com.nonpolynomial.btleplug.android.impl.Peripheral.asyncWithFuture(Peripheral.java:324)
                                                                                                    	at com.nonpolynomial.btleplug.android.impl.Peripheral.access$700(Peripheral.java:24)
                                                                                                    	at com.nonpolynomial.btleplug.android.impl.Peripheral$6.onDescriptorWrite(Peripheral.java:244)
                                                                                                    	at com.nonpolynomial.btleplug.android.impl.Peripheral$Callback.onDescriptorWrite(Peripheral.java:408)
                                                                                                    	at android.bluetooth.BluetoothGatt$1$10.run(BluetoothGatt.java:636)
                                                                                                    	at android.bluetooth.BluetoothGatt.runOrQueueCallback(BluetoothGatt.java:864)
                                                                                                    	at android.bluetooth.BluetoothGatt.-$$Nest$mrunOrQueueCallback(Unknown Source:0)
                                                                                                    	at android.bluetooth.BluetoothGatt$1.onDescriptorWrite(BluetoothGatt.java:631)
                                                                                                    	at android.bluetooth.IBluetoothGattCallback$Stub.onTransact(IBluetoothGattCallback.java:234)
                                                                                                    	at android.os.Binder.execTransactInternal(Binder.java:1285)
                                                                                                    	at android.os.Binder.execTransact(Binder.java:1244)`

--- a/src/droidplug/peripheral.rs
+++ b/src/droidplug/peripheral.rs
@@ -47,8 +47,11 @@ fn get_poll_result<'a: 'b, 'b>(
     env: &'b JNIEnv<'a>,
     result: JPollResult<'a, 'b>,
 ) -> Result<JObject<'a>> {
+    let future_exc =
+        jni_utils::classcache::get_class("io/github/gedgygedgy/rust/future/FutureException")
+            .unwrap();
     try_block(env, || Ok(Ok(result.get()?)))
-        .catch("io/github/gedgygedgy/rust/future/FutureException", |ex| {
+        .catch(future_exc.as_obj(), |ex| {
             let cause = env
                 .call_method(ex, "getCause", "()Ljava/lang/Throwable;", &[])?
                 .l()?;

